### PR TITLE
Support symlinks in the upload directory

### LINKED ISSUE #4138: Backend  Unable to list contents from 'files'.
### Affected version(s)

4.13.0-RC-3

### Description

Der Aufruf der Syncronisation in der Dateiverwaltung führt in der aktuellen Contao Version 4.13.0-rc-3 zu diesem Fehler.
```
Contao\CoreBundle\Filesystem\VirtualFilesystemException:
Unable to list contents from 'files'.

  at vendor/contao/core-bundle/src/Filesystem/VirtualFilesystemException.php:138
  at Contao\CoreBundle\Filesystem\VirtualFilesystemException::unableToListContents('files', object(SymbolicLinkEncountered))
     (vendor/contao/core-bundle/src/Filesystem/MountManager.php:417)
  at Contao\CoreBundle\Filesystem\MountManager->doListContents('files', false)
     (vendor/contao/core-bundle/src/Filesystem/MountManager.php:316)
  at Contao\CoreBundle\Filesystem\MountManager->listContents('files', false)
     (vendor/contao/core-bundle/src/Filesystem/VirtualFilesystem.php:308)
  at Contao\CoreBundle\Filesystem\VirtualFilesystem->doListContents('files/layout', false, 1)
     (vendor/contao/core-bundle/src/Filesystem/FilesystemItemIterator.php:64)
  at Contao\CoreBundle\Filesystem\FilesystemItemIterator->getIterator()
     (vendor/contao/core-bundle/src/Filesystem/Dbafs/Dbafs.php:722)
  at Contao\CoreBundle\Filesystem\Dbafs\Dbafs->Contao\CoreBundle\Filesystem\Dbafs\{closure}('')
     (vendor/contao/core-bundle/src/Filesystem/Dbafs/Dbafs.php:772)
  at Contao\CoreBundle\Filesystem\Dbafs\Dbafs->getFilesystemPaths(array(''), array())
     (vendor/contao/core-bundle/src/Filesystem/Dbafs/Dbafs.php:317)
 at Contao\CoreBundle\Filesystem\Dbafs\Dbafs->doComputeChangeSet(array('layout' => 1, 'published' => 1, 'published/images' => 1, 'published/images/logo.png' => 0, 'published/download' => 1, 'layout/src' => 1, 'layout/src/scss' => 1, 'layout/src/scss/default.scss' => 0, 'published/uploads' => 1, .... )
     (vendor/contao/core-bundle/src/Filesystem/Dbafs/Dbafs.php:219)
  at Contao\CoreBundle\Filesystem\Dbafs\Dbafs->sync()
     (vendor/contao/core-bundle/src/Filesystem/Dbafs/DbafsManager.php:273)
  at Contao\CoreBundle\Filesystem\Dbafs\DbafsManager->sync()
     (vendor/contao/core-bundle/src/Resources/contao/drivers/DC_Folder.php:2486)
  at Contao\DC_Folder->sync()
     (vendor/contao/core-bundle/src/Resources/contao/classes/Backend.php:655)
  at Contao\Backend->getBackendModule('files', null)
     (vendor/contao/core-bundle/src/Resources/contao/controllers/BackendMain.php:170)
  at Contao\BackendMain->run()
     (vendor/contao/core-bundle/src/Controller/BackendController.php:49)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (vendor/symfony/http-kernel/HttpKernel.php:152)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (vendor/symfony/http-kernel/HttpKernel.php:74)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (vendor/symfony/http-kernel/Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (public/index.php:44) 
```

### PR DESCRIPTION
**Pull request by m-vo**
Fixes #4138

This PR adds a BC layer for symlinks in the upload directory: 

* We now skip symlinks in our Flysystem local filesystem adapters (= no exception will be thrown when they are encountered).
* We mount a local adapter for each found symlink in the `ConfigureFilesystemPass` compiler pass.

Todo:
- [x]  Add tests

### CODE REVIEW

**Comment by m-vo** [comment]
@Ainschy Can you test if the changes from this PR fix your issue? Make sure to rebuild the container cache. :slightly_smiling_face:

**Review by ausi — DISMISSED** [review]

**Comment by Ainschy** [comment]
Works fine for me. Thx @m-vo

**Comment by m-vo** [comment]
There is still a failing unit test for the `symlink relative to the files directory` case on Windows, but it's too hard for me to debug this by force pushing commit after commit to the CI. 

I would either skip the test on windows or maybe @fritzmg might want to check this out locally and see if we can fix it. :see_no_evil:

**Comment by leofeyer** [comment]
We obviously cannot leave it broken on Windows. 😕

**Comment by Ainschy** [comment]
i'm working allso under windows 10, all looks fine by symlinks.

**Comment by m-vo** [comment]
> We obviously cannot leave it broken on Windows. confused

It's most likely only the test that fails, not the implementation. Plus: Judging by how cumbersome it is to create a relative symlink on windows, I'm pretty sure nobody is using this. :smile:

**Comment by leofeyer** [comment]
Yeah, I think I recall that Windows only supports absolute symlinks. Maybe @fritzmg can help here.

**Comment by fritzmg** [comment]
> Plus: Judging by how cumbersome it is to create a relative symlink on windows, I'm pretty sure nobody is using this.

Creating relative symlinks on Windows is not cumbersome, it works the same way as in Linux. It's just that PHP does not support it.

**Comment by leofeyer** [comment]
Ok, then we should probably skip the test on Windows.

**Comment by m-vo** [comment]
> Creating relative symlinks on Windows is not cumbersome, it works the same way as in Linux. It's just that PHP does not support it.

That's what I was talking about. 🙃

**Comment by m-vo** [comment]
Thanks to @fritzmg and @ausi we got the testsuite green again. And it also works on Windows now. :muscle: 

And in case you're interested, that's how you can create a relative symlink cross platform:
```php
private function createSymlink(string $target, string $link, string $cwd): void
{
    if ('\\' === \DIRECTORY_SEPARATOR) {
        $target = str_replace('/', '\\', $target);
        $link = str_replace('/', '\\', $link);

        Process::fromShellCommandline('mklink /d "${:link}" "${:target}"', $cwd)
            ->mustRun(null, ['link' => $link, 'target' => $target])
        ;
    } else {
        chdir($cwd);

        /** @phpstan-ignore-next-line because we need to create relative symlinks and cannot use the Symfony Filesystem for that */
        symlink($target, $link);
    }
}
```

**Review by ausi — DISMISSED** [review]

File: core-bundle/tests/DependencyInjection/Compiler/ConfigureFilesystemPassTest.php [hunk 1]
```diff
@@ -57,4 +72,86 @@ function (FilesystemConfiguration $config) use ($container): bool {
...
+    {
+        if ('\\' === \DIRECTORY_SEPARATOR) {
+            $target = str_replace('/', '\\', $target);
+            $link = str_replace('/', '\\', $link);
+
+            Process::fromShellCommandline('mklink /d "${:link}" "${:target}"', $cwd)
+                ->mustRun(null, ['link' => $link, 'target' => $target])
+            ;
+        } else {
+            chdir($cwd);
```
Comment (ausi) [inline core-bundle/tests/DependencyInjection/Compiler/ConfigureFilesystemPassTest.php:176]: This should get reset in the tearDown() method 🤓

Comment (m-vo) [reply core-bundle/tests/DependencyInjection/Compiler/ConfigureFilesystemPassTest.php:176]: Should we record the current path with `getcwd` or is there a default we could use?

Comment (ausi) [reply core-bundle/tests/DependencyInjection/Compiler/ConfigureFilesystemPassTest.php:176]: I think recording is the way to go…

Comment (m-vo) [reply core-bundle/tests/DependencyInjection/Compiler/ConfigureFilesystemPassTest.php:176]: implemented in a626f31

**Comment by fritzmg** [comment]
There are still some issues:

* Symlinks to files do not work.
* Symlinks within other symlinks do not work.

The latter did not work before either though. Not sure how important it is to support the former.

**Comment by m-vo** [comment]
> Symlinks to files do not work.
> Symlinks within other symlinks do not work.

I don't think we can handle single files. What would be a use case for that?

And I would honestly ignore nested symlinks, that seems too niche to me and has its own problems anyways (for example `realpath` does not resolve more than one layer on Windows).

**Review by ausi — DISMISSED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @m-vo.

**Pull request merged**